### DISCUSSION TITLE
Resolve Multiple Receptions correctly

### DISCUSSION THREAD

**Discussion by TimonPetermann** [inet-framework/inet#632] [Ideas]
Hey guys,

I am currently working on a software framework based on OMNET and INET for satellite applications.
We aim to predict satellite contacts to calculate contact and interference plans.
At the moment, we observe the behavior, that at any time, where a second reception is resolved at the receiver, the reception is not attempted, what results into a packet loss.
Nevertheless, this should not be the case if the second transmission is received with much more power right?
Then, the first reception should be discarded, not the second.
Is this an unintended behavior of the implementation or am I using something wrong?
Or if the behavior is indended, what is the reason for this?

I think the issue is the `ReceiverBase::computeIsReceptionAttempted` function call, which returns false due to the reception that is already going on:
```c++
auto transmission = reception->getReceiver()->getReceptionInProgress();
return transmission == nullptr || transmission == reception->getTransmission();
```

In the given case, the reception meets all SNIR, BW, etc. requirements, as the receptionPossible parameter returns true.
The used receiver is the `ApskScalarReceiver` and we are working on inet 4.2.

I'm looking forward to discuss this here, thank you!

**Comment by levy** [comment]
Indeed, INET currently doesn't support capturing a stronger signal when a weaker signal is already being received. This change would not be too difficult though, because all information is already present in the physical layer and transmission medium. Probably something like a threshold parameter would make it work that specifies how stronger the to be captured signal must be compared to the one that is currently being received.

Christmas is here and the year is ending, but we can come back to this issue in January and make it work.

**Reply by TimonPetermann** [reply]
I wish a happy new year!

Okay, I am not quite sure if a threshold is a good solution, as usually the SNIR threshold is the relevant parameter in this case.
I would suggest to filter for the strongest signal and then (re-)calculated the SNIR with other on-going receptions as interference.
Then, the SNIR calculation has to be done every time a new reception arrives with all on-going receptions.
The weaker signals are marked then as collided.

 
I would offer to also have a look at this, I think I can find some time for it in the next weeks.

**Comment by levy** [comment]
There are a few very different decisions in the receiver at play here. The following decisions are done by the receiver at different points in the reception process:
- Is the reception of the signal possible? For example, a differently modulated signal cannot be received independently of reception conditions. This decision is made when the **start** of the transmitted signal **departs from the transmitter** on the transmission medium.
- Is the reception of the signal attempted? For example, a too weak signal cannot be received independently of the SNIR. This decision is made when the **start** of the received signal **arrives at the receiver** on the transmission medium, so it can use the signal power because it's already known, but it cannot use the SNIR for the total duration of the signal, because it is not yet fully known (e.g. at any future moment a new signal may enter the transmission medium at any distance from the receiver).
- Is the reception of the signal successful? For example, a signal having too low SNIR cannot be received. This decision is made when the **end** of the received signal **arrives at the receiver** on the transmission medium, so it can use the SNIR for the total duration of the signal, because it is already known.

Now, for signal capture I think we need to introduce a new decision into the reception process:
- Is the capture of the second signal attempted? This decision would be made when the **start** of the second signal **arrives at the receiver** on the transmission medium. Once the decision is made to switch over to the second signal, the reception process would be the same as if the first signal reception would not have been attempted at all. The big question is what should the capture decision be based on?

Here's an example:
 - noise power is 10
 - first signal's power is 100
 - second signal's power is 101

Should the receiver really switch over to the second signal? I understand that the second signal is stronger than the first, both in absolute power and in SNIR at the start of the reception of the second, but is it really what would happen in a real world receiver? Obviously if the second signal's power is like 1000, the receiver should capture it. But what if the receiver can detect if a signal is arriving only when the SNIR is greater than 10 (because otherwise the signal is just treated as high level noise)? Then the reception attempt of the first signal is correct because it's SNIR is 10 at its start, but for the second signal the SNIR is less than 1 at its own start. Even though the SNIR at that point for the first signal is also less than 1 (even more so than for the second), but how does the receiver distinguish this case from the arrival of different signal which would not even pass the first receiver check, namely the one that checks if the signal cannot be received due to modulation differences. So, in this case I would expect the receiver to stick with the reception process of the first signal and fail. Moreover, the fact that the receiver did not attempt the reception of the second signal may affect the reception of future signals that may or may not interfere with the first and/or second signals, etc.

This gets kinda complicated but I feel that simply finding the strongest signal is not right because the real reception process doesn't work like that.

**Reply by rhornig** [reply]
IMHO, this feature is meant to increase the throughput of the radio medium. The receiver should switch to receive the second packet only if the estimated possibility of receiving the second packet correctly (with the ongoing first packet treated as noise) is **considerably** higher than the probability of receiving the first packet correctly. I think having examples like the above (with the almost same power) is not useful as there is no good decision in these cases. The reception will surely fail either way, no mater if we switch or not. 

The interesting case happens when the second incoming signal is so much stronger than the first one, that even if we would treat the first packet as a noise, the second packet still has a reasonable chance for reception. The only data we know at this point is the power of the background noise, the first (receiving) signal, and the second (arriving) signal. So a reasonable assumption would be to switch to the reception of the second signal only if the second signal's SNR (including the first signal as noise) is above a certain (preconfigured) level. The actual optional value of this SNR is to be debated, but it could be determined by running an experiment (it may or may not depend on modulation parameters and error correction).

In short, I'm also backing the idea, to use a simple SNR threshold.

**Reply by TimonPetermann** [reply]
I also think in this example it does not matter which signal is attempted, as a real receiver probably will not be able to track any on the signals. (So none of them should be attempted anymore? -> Is a not attempted signal still accessible for potential interference calculation if e.g. a third signal arrives?)

I am not sure if I get rhornig right, but the SNIR parameter we need in this case should be the SNIR value that already exists in the receiver. Or what is meant by the optional SNR value, do you propose to add an additional SNR value?

**Reply by edlongman** [reply]
I agree with the comment that real receivers probably won't be able to track any of the signals. 
If a signal is being recieved and a second signal starts being with recieved above the noise threshold it should discard both signals. Which I think is what it does anyway?

**Reply by levy** [reply]
Right now the first reception will be attempted and will fail, the second reception will not even be attempted.
